cpp/src/strings/replace/replace.cu

```diff
@@ -0,0 +1,417 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/strings/replace.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include "../utilities.hpp"
+#include "../utilities.cuh"
+
+namespace
+{
+
+/**
+ * @brief Used as template parameter to divide size calculation from
+ * the actual string operation within a function.
+ *
+ * Useful when most of the logic is identical for both passes.
+ */
+enum TwoPass
+{
+    SizeOnly = 0, ///< calculate the size only
+    ExecuteOp     ///< run the string operation
+};
+
+} // namespace
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+namespace
+{
+
+/**
+ * @brief Function logic for the replace API.
+ *
+ * This will perform a replace operation on each string.
+ */
+template <TwoPass Pass=SizeOnly>
+struct replace_fn
+{
+    column_device_view const d_strings;
+    string_view const d_target;
+    string_view const d_repl;
+    int32_t max_repl;
+    const int32_t* d_offsets{};
+    char* d_chars{};
+
+    __device__ size_type operator()(size_type idx)
+    {
+        if( d_strings.is_null(idx) )
+            return 0; // null string
+        string_view d_str = d_strings.element<string_view>(idx);
+        auto length = d_str.length();
+        auto max_n = max_repl;
+        if( max_n < 0 )
+            max_n = length; // max possible replacements
+        char* out_ptr = nullptr;
+        if( Pass==ExecuteOp )
+            out_ptr = d_chars + d_offsets[idx];
+        const char* in_ptr = d_str.data();
+        size_type bytes = d_str.size_bytes();
+        auto position = d_str.find(d_target);
+        size_type last_pos = 0;
+        while( (position >= 0) && (max_n > 0) )
+        {
+            if( Pass==SizeOnly )
+                bytes += d_repl.size_bytes() - d_target.size_bytes();
+            else // ExecuteOp
+            {
+                size_type curr_pos = d_str.byte_offset(position);
+                out_ptr = copy_and_increment(out_ptr, in_ptr + last_pos, curr_pos - last_pos); // copy left
+                out_ptr = copy_string(out_ptr, d_repl); // copy repl
+                last_pos = curr_pos + d_target.size_bytes();
+            }
+            position = d_str.find(d_target, position + d_target.size_bytes() );
+            --max_n;
+        }
+        if( Pass==ExecuteOp ) // copy whats left (or right depending on your point of view)
+            copy_and_increment(out_ptr, in_ptr + last_pos, d_str.size_bytes() - last_pos);
```

Comment:
copy_and_increment is overkill here, since you don't use its output (the increment). Just call `memcpy` since it will be clearer.